service/src/main/java/uk/gov/hmcts/reform/fpl/service/ApplicationDocumentsService.java

```diff
@@ -0,0 +1,96 @@
+
+        Map<String, Object> updatedCaseData = new HashMap<>();
+
+        updatedCaseData.put("documents", updatedDocuments);
+
+        return updatedCaseData;
+    }
+
+    private List<Element<ApplicationDocument>> setUpdatedByAndDateAndTimeOnDocuments(
+        List<Element<ApplicationDocument>> currentDocuments,
+        List<Element<ApplicationDocument>> previousDocuments) {
+
+        if (isEmpty(previousDocuments) && !isEmpty(currentDocuments)) {
+            currentDocuments.stream().forEach(this::setUpdatedByAndDateAndTimeOnDocumentToCurrent);
+            return currentDocuments;
+        } else {
+            return currentDocuments.stream()
+                .map(document -> {
+                    Optional<Element<ApplicationDocument>> documentBefore = getDocumentBeforeFromID(document.getId(),
+                        previousDocuments);
+
+                    if (documentBefore.isPresent()) {
+                        Element<ApplicationDocument> oldDocument = documentBefore.get();
+                        handleExistingDocuments(oldDocument, document);
+
+                    } else {
+                        // New document was added
+                        setUpdatedByAndDateAndTimeOnDocumentToCurrent(document);
+                    }
+                    return document;
+                }).collect(Collectors.toList());
+        }
+    }
+
+    private void handleExistingDocuments(Element<ApplicationDocument> documentBefore,
+                                         Element<ApplicationDocument> document) {
+        if (documentBefore.getId().equals(document.getId())) {
+            if (documentBefore.getValue().getDocument().equals(document.getValue().getDocument())) {
+                // Document wasn't modified so persist old values
+                document.getValue().setDateTimeUploaded(documentBefore.getValue().getDateTimeUploaded());
+                document.getValue().setUploadedBy(documentBefore.getValue().getUploadedBy());
+            } else {
+                // Document was modified so update
+                setUpdatedByAndDateAndTimeOnDocumentToCurrent(document);
+            }
+        }
+    }
+
+    private Optional<Element<ApplicationDocument>> getDocumentBeforeFromID(UUID documentID,
+                                             List<Element<ApplicationDocument>> previousDocuments) {
+        return Optional.ofNullable(previousDocuments.stream()
+            .filter(previousDocument -> previousDocument.getId()
+            .equals(documentID))
+            .findAny()
+            .orElse(null));
+    }
+
+    private Element<ApplicationDocument> setUpdatedByAndDateAndTimeOnDocumentToCurrent(
+        Element<ApplicationDocument> document) {
+        String uploadedBy = documentUploadHelper.getUploadedDocumentUserDetails();
+
+        document.getValue().setDateTimeUploaded(time.now());
+        document.getValue().setUploadedBy(uploadedBy);
+        return document;
```

Comment:
I think it's just as readable having this method in the logic above rather than method.

Reply:
discussed